---
title: "Impact of dysfunctional ubiquitination in response to cancer immunotherapy"
subtitle: <center> Somatic mutations analysis - Gene Overrepresentation Analysis </center>
author: "Eric Bautista Farrerons (s212514)"
date: '`r paste("First created on January 2024. Updated on ", format(Sys.Date(), "%d %B %Y"))`'
output:
  html_document:
    css: style.css
    code_folding: hide
    fig_caption: yes
    toc: yes
    toc_float:
      collapsed: true
      smooth_scroll: true
    toc_depth: 4
    number_sections: true
---

# Load data

```{r, message=FALSE}
### Load required libraries
library(readxl)
library(tidyverse)
library(purrr)
library(dplyr)
library(magrittr)
library(knitr)
library(DT)
library(httr)
library(jsonlite)
library(stringr)
library(GSEABase)

### Load functions
source(file = "../02_functions.R")

### Define paths
current_dir <- getwd()
data_path <- file.path(current_dir,
                       "../../data/_raw")
interesting_gs_path <- file.path(data_path,
                                 "../interesting_genesets.xlsx")
results_pc_path <- file.path(current_dir,
                              "../../results/patient_characteristics/output")
results_fip_path <- file.path(current_dir,
                              "../../results/somatic_mutation_analysis/functional_impact_prediction/output")
results_gor_path <- file.path(current_dir, 
                                 "../../results/somatic_mutation_analysis/gene_overrepresentation_analysis/output")

# Create the folder if it does not exist
if(!file.exists(results_gor_path)) {
  dir.create(results_gor_path, recursive = TRUE)
}

figures_gor_path <- file.path(current_dir, 
                                 "../../results/somatic_mutation_analysis/gene_overrepresentation_analysis/figures")

# Create the folder if it does not exist
if(!file.exists(figures_gor_path)) {
  dir.create(figures_gor_path, recursive = TRUE)
}

maf_df_path <- file.path(results_fip_path, "all_mutations.csv") # All mutations df
nonsyn_df_path <- file.path(results_fip_path, "nonsyn_mutations.csv") # Nonsyn mutations df
lof_df_path <- file.path(results_fip_path, "lof_mutations_reduced.csv") # LoF mutations df

# Sup1 data
sup1_response_path <- file.path(results_pc_path, 
                           "sup1_response.csv")
### Read data & wrangle

# Read 3 lists of mutations

# All
maf_df_annotated <- read.csv(maf_df_path, header = TRUE)[ , -1]
maf_df_annotated <- maf_df_annotated %>% 
  dplyr::filter(!(sample_id %in% c("MM909_20", "MM909_24")))
list_of_all_mutations <- split(maf_df_annotated, maf_df_annotated$sample_id
                               )
# Nonsyn
nonsyn_df_annotated <- read.csv(nonsyn_df_path, header = TRUE)[ , -1]
nonsyn_df_annotated <- nonsyn_df_annotated %>% 
  dplyr::filter(!(sample_id %in% c("MM909_20", "MM909_24")))
list_of_nonsyn_mutations <- split(nonsyn_df_annotated, nonsyn_df_annotated$sample_id)

# LoF
lof_df_annotated <- read.csv(lof_df_path, header = TRUE)[ , -1]
lof_df_annotated <- lof_df_annotated %>% 
  dplyr::filter(!(sample_id %in% c("MM909_20", "MM909_24")))
list_of_lof_mutations <- split(lof_df_annotated, lof_df_annotated$sample_id)

# Get the list of patient IDs
sample_id_list <- unique(maf_df_annotated$sample_id)

## Read sup1_response.csv
sup1_response_df <- read.csv(sup1_response_path)[ , -1] # Drop first column

# df to merge response info later
response_df <- sup1_response_df %>% 
  dplyr::rename(sample_id = Sample.ID,
                patient_response = Patient.Response) %>% 
  dplyr::select(sample_id, patient_response)

# Create a list of Responders and Non-responders
sample_id_lists <- split(sup1_response_df$Sample.ID, 
                         sup1_response_df$Patient.Response) # Split the "Sample.ID" values into lists based on "Patient.Response"

r_sample_ids <- sample_id_lists$R
nr_sample_ids <- sample_id_lists$NR

## Read the gene sets Excel file
interesting_gs <- read_excel(interesting_gs_path)
interesting_gs_list <- na.omit(as.character(interesting_gs[["Gene Set"]])) # Extract the "Gene Set" column, convert to character and omit NAs
boolean_gs_list <- paste(interesting_gs_list, collapse = " OR ") # This is to generate the query to search in MSigDB
```

# Summary of mutations per patient

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
# Responders sum plot

# Define output path
mm909_sum_plot_path <- file.path(figures_gor_path,
                              "mm909_sum_plot_r.png")

# Summary plot
mm909_sum_plot <- sum_plot(response_df, maf_df_annotated, nonsyn_df_annotated, lof_df_annotated, mm909_sum_plot_path, select_r = TRUE)

# Display
print(mm909_sum_plot)
```

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=8}
# Non-responders sum plot

# Define output path
mm909_sum_plot_path <- file.path(figures_gor_path,
                              "mm909_sum_plot_nr.png")

# Summary plot
mm909_sum_plot <- sum_plot(response_df, maf_df_annotated, nonsyn_df_annotated, lof_df_annotated, mm909_sum_plot_path, select_r = FALSE)

# Display
print(mm909_sum_plot)
```

# Pathway analysis (gene overrepresentation)

## First analysis using C5 - BP

Consists of gene sets derived from the Gene Ontology (GO) annotations. Here, we focus on Gene sets derived from the GO Biological Process (BP) ontology.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_c5bp_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_c5bp_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_c5bp_lof.csv")

# # Perform FORA analysis for C5:BP Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, category="C5", subcategory="BP")

# Load fora results
fora_results_c5bp_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_c5bp_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_c5bp_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top C5 BP across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_c5bp_all_plot_path <- file.path(figures_gor_path,
                              "mm909_c5bp_all_plot.png")

mm909_c5bp_all_plot <- plot_top5(response_df, fora_results_c5bp_all_df, mm909_c5bp_all_plot_path, "FORA with top C5 BP across samples (All mutations)", "Top C5 BP Gene Sets")

print(mm909_c5bp_all_plot)
```

### Visualization: Top C5 BP across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_c5bp_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_c5bp_nonsyn_plot.png")

mm909_c5bp_nonsyn_plot <- plot_top5(response_df, fora_results_c5bp_nonsyn_df, mm909_c5bp_nonsyn_plot_path, "FORA with top C5 BP across samples (Nonsyn mutations)", "Top C5 BP Gene Sets")

print(mm909_c5bp_nonsyn_plot)
```

### Visualization: Top C5 BP across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_c5bp_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_c5bp_lof_plot.png")

mm909_c5bp_lof_plot <- plot_top5(response_df, fora_results_c5bp_lof_df, mm909_c5bp_lof_plot_path, "FORA with top C5 BP across samples (LoF mutations)", "Top C5 BP Gene Sets")

print(mm909_c5bp_lof_plot)
```

## Second analysis using C5 - CC

Consists of gene sets derived from the Gene Ontology (GO) annotations. Here, we focus on Gene sets derived from the GO Cellular Component (CC) ontology.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_c5cc_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_c5cc_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_c5cc_lof.csv")

# # Perform FORA analysis for CC Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, category="C5", subcategory="CC")

# Load fora results
fora_results_c5cc_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_c5cc_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_c5cc_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: C5 CC PROTEASOME across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=5}
# Define output path
mm909_c5cc_proteasome_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_c5cc_proteasome_lof_plot.png")

proteasome_genesets = c("GOCC_PROTEASOME_ACCESSORY_COMPLEX", "GOCC_PROTEASOME_COMPLEX", "GOCC_PROTEASOME_CORE_COMPLEX", "GOCC_PROTEASOME_CORE_COMPLEX_ALPHA_SUBUNIT_COMPLEX", "GOCC_PROTEASOME_CORE_COMPLEX_BETA_SUBUNIT_COMPLEX", "GOCC_PROTEASOME_REGULATORY_PARTICLE", "GOCC_PROTEASOME_REGULATORY_PARTICLE_BASE_SUBCOMPLEX", "GOCC_PROTEASOME_REGULATORY_PARTICLE_LID_SUBCOMPLEX")

mm909_c5cc_proteasome_lof_plot <- plot_geneset(proteasome_genesets, response_df, fora_results_c5cc_lof_df, mm909_c5cc_proteasome_lof_plot_path, "FORA with PROTEASOME-related C5 CC Gene Sets across samples (LoF mutations)", "PROTEASOME Gene Sets")

print(mm909_c5cc_proteasome_lof_plot)
```

## Third analysis using H

Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_h_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_h_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_h_lof.csv")

# # Perform FORA analysis for H Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, category="H")

# Load fora results
fora_results_h_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_h_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_h_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top H across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
# Define output path
mm909_h_all_plot_path <- file.path(figures_gor_path,
                              "mm909_h_all_plot.png")

mm909_h_all_plot <- plot_top5(response_df, fora_results_h_all_df, mm909_h_all_plot_path, "FORA with top H across samples (All mutations)", "Top H Gene Sets")

print(mm909_h_all_plot)
```

### Visualization: Top H across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
# Define output path
mm909_h_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_h_nonsyn_plot.png")

mm909_h_nonsyn_plot <- plot_top5(response_df, fora_results_h_nonsyn_df, mm909_h_nonsyn_plot_path, "FORA with top H across samples (Nonsyn mutations)", "Top H Gene Sets")

print(mm909_h_nonsyn_plot)
```

### Visualization: Top H across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
# Define output path
mm909_h_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_h_lof_plot.png")

mm909_h_lof_plot <- plot_top5(response_df, fora_results_h_lof_df, mm909_h_lof_plot_path, "FORA with top H across samples (LoF mutations)", "Top H Gene Sets")

print(mm909_h_lof_plot)
```

### Visualization: HALLMARK_APOPTOSIS across samples stratified by response (LoF)

Filter specifically for HALLMARK_APOPTOSIS gene set.

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=5}
# Define output path
mm909_h_apoptosis_plot_path <- file.path(figures_gor_path,
                              "mm909_h_apoptosis_plot.png")

mm909_h_apoptosis_plot <- plot_geneset(c("HALLMARK_APOPTOSIS"), response_df, fora_results_h_lof_df, mm909_h_apoptosis_plot_path, "FORA with HALLMARK_APOPTOSIS across samples (LoF mutations)", "HALLMARK_APOPTOSIS Gene Set")

print(mm909_h_apoptosis_plot)
```

## Fourth analysis using C6 (oncogenic)

Gene sets that represent signatures of cellular pathways which are often dis-regulated in cancer. The majority of signatures were generated directly from microarray data from NCBI GEO or from internal unpublished profiling experiments involving perturbation of known cancer genes.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_c6_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_c6_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_c6_lof.csv")

# # Perform FORA analysis for C6 Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, category="C6")

# Load fora results
fora_results_c6_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_c6_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_c6_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top C6 across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=15}
# Define output path
mm909_c6_all_plot_path <- file.path(figures_gor_path,
                              "mm909_c6_all_plot.png")

mm909_c6_all_plot <- plot_top5(response_df, fora_results_c6_all_df, mm909_c6_all_plot_path, "FORA with top C6 across samples (All mutations)", "Top C6 Gene Sets")

print(mm909_c6_all_plot)
```

### Visualization: Top C6 across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=15}
# Define output path
mm909_c6_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_c6_nonsyn_plot.png")

mm909_c6_nonsyn_plot <- plot_top5(response_df, fora_results_c6_nonsyn_df, mm909_c6_nonsyn_plot_path, "FORA with top C6 across samples (Nonsyn mutations)", "Top C6 Gene Sets")

print(mm909_c6_nonsyn_plot)
```

### Visualization: Top C6 across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=15}
# Define output path
mm909_c6_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_c6_lof_plot.png")

mm909_c6_lof_plot <- plot_top5(response_df, fora_results_c6_lof_df, mm909_c6_lof_plot_path, "FORA with top C6 across samples (LoF mutations)", "Top C6 Gene Sets")

print(mm909_c6_lof_plot)
```

## Fifth analysis using C7 (immunologic)

Gene sets that represent cell states and perturbations within the immune system.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_c7_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_c7_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_c7_lof.csv")

# # Perform FORA analysis for C7 Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, category="C7")

# Load fora results
fora_results_c7_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_c7_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_c7_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top C7 across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=20}
# Define output path
mm909_c7_all_plot_path <- file.path(figures_gor_path,
                              "mm909_c7_all_plot.png")

mm909_c7_all_plot <- plot_top5(response_df, fora_results_c7_all_df, mm909_c7_all_plot_path, "FORA with top C7 across samples (All mutations)", "Top C7 Gene Sets")

print(mm909_c7_all_plot)
```

### Visualization: Top C7 across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=25}
# Define output path
mm909_c7_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_c7_nonsyn_plot.png")

mm909_c7_nonsyn_plot <- plot_top5(response_df, fora_results_c7_nonsyn_df, mm909_c7_nonsyn_plot_path, "FORA with top C7 across samples (Nonsyn mutations)", "Top C7 Gene Sets")

print(mm909_c7_nonsyn_plot)
```

### Visualization: Top C7 across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=25}
# Define output path
mm909_c7_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_c7_lof_plot.png")

mm909_c7_lof_plot <- plot_top5(response_df, fora_results_c7_lof_df, mm909_c7_lof_plot_path, "FORA with top C7 across samples (LoF mutations)", "Top C7 Gene Sets")

print(mm909_c7_lof_plot)
```

## Sixth analysis using MSigDB filter of "Ubiquitin"

I searched for the term "ubiquitin" in MSigDB and downloaded a GMT file with all the hits (Gene Sets). Number of GS: 101.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_ubiquitin_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_ubiquitin_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_ubiquitin_lof.csv")

# # Load GMT file as a GeneSetCollection object
# gene_sets_collection <- getGmt(file.path(data_path, "genesets_ubiquitin.v2023.2.Hs.gmt"))
# 
# # Initialize an empty list to store the results
# gene_sets_list <- list()
# 
# # Iterate over each GeneSet in the GeneSetCollection
# for (setName in names(gene_sets_collection)) {
#   # Extract the GeneSet by its name
#   geneSet <- gene_sets_collection[[setName]]
# 
#   # Convert Hugo symbols to Ensembl IDs for this GeneSet
#   ensembl_ids <- SYMBOLtoENSEMBL(geneIds(geneSet))
# 
#   # Ensure the result is a character vector of Ensembl Gene IDs
#   ensembl_ids_vector <- as.character(ensembl_ids)
# 
#   # Store the converted IDs in the list, associated with the setName
#   gene_sets_list[[setName]] <- ensembl_ids_vector
# }
# 
# # Perform FORA analysis for Ubiquitin Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, gene_sets_list=gene_sets_list)

# Load fora results
fora_results_ubiquitin_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_ubiquitin_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_ubiquitin_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top Ubiquitin GS across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}
# Define output path
mm909_ubiquitin_all_plot_path <- file.path(figures_gor_path,
                              "mm909_ubiquitin_all_plot.png")

mm909_ubiquitin_all_plot <- plot_top5(response_df, fora_results_ubiquitin_all_df, mm909_ubiquitin_all_plot_path, "FORA with top Ubiquitin GS across samples (All mutations)", "Top Ubiquitin Gene Sets")

print(mm909_ubiquitin_all_plot)
```

### Visualization: Top Ubiquitin GS across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}
# Define output path
mm909_ubiquitin_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_ubiquitin_nonsyn_plot.png")

mm909_ubiquitin_nonsyn_plot <- plot_top5(response_df, fora_results_ubiquitin_nonsyn_df, mm909_ubiquitin_nonsyn_plot_path, "FORA with top Ubiquitin GS across samples (Nonsyn mutations)", "Top Ubiquitin Gene Sets")

print(mm909_ubiquitin_nonsyn_plot)
```

### Visualization: Top Ubiquitin GS across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}
# Define output path
mm909_ubiquitin_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_ubiquitin_lof_plot.png")

mm909_ubiquitin_lof_plot <- plot_top5(response_df, fora_results_ubiquitin_lof_df, mm909_ubiquitin_lof_plot_path, "FORA with top Ubiquitin GS across samples (LoF mutations)", "Top Ubiquitin Gene Sets")

print(mm909_ubiquitin_lof_plot)
```

## Seventh analysis using C2:CGP (chemical & genetic perturbations)

Gene sets in this collection are curated from various sources, including online pathway databases and the biomedical literature. Many sets are also contributed by individual domain experts. Here, we focus on Chemical and genetic perturbations (CGP).

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_c2cgp_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_c2cgp_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_c2cgp_lof.csv")

# # Perform FORA analysis for C2_CGP Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, category="C2", subcategory="CGP")

# Load fora results
fora_results_c2cgp_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_c2cgp_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_c2cgp_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top C2:CGP across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_c2cgp_all_plot_path <- file.path(figures_gor_path,
                              "mm909_c2cgp_all_plot.png")

mm909_c2cgp_all_plot <- plot_top5(response_df, fora_results_c2cgp_all_df, mm909_c2cgp_all_plot_path, "FORA with top C2:CGP across samples (All mutations)", "Top C2:CGP Gene Sets")

print(mm909_c2cgp_all_plot)
```

### Visualization: Top C2:CGP across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_c2cgp_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_c2cgp_nonsyn_plot.png")

mm909_c2cgp_nonsyn_plot <- plot_top5(response_df, fora_results_c2cgp_nonsyn_df, mm909_c2cgp_nonsyn_plot_path, "FORA with top C2:CGP across samples (Nonsyn mutations)", "Top C2:CGP Gene Sets")

print(mm909_c2cgp_nonsyn_plot)
```

### Visualization: Top C2:CGP across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_c2cgp_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_c2cgp_lof_plot.png")

mm909_c2cgp_lof_plot <- plot_top5(response_df, fora_results_c2cgp_lof_df, mm909_c2cgp_lof_plot_path, "FORA with top C2:CGP across samples (LoF mutations)", "Top C2:CGP Gene Sets")

print(mm909_c2cgp_lof_plot)
```

## Eighth analysis using C2:CP:KEGG_MEDICUS

Gene sets in this collection are curated from various sources, including online pathway databases and the biomedical literature. Many sets are also contributed by individual domain experts. Here, we focus on Canonical pathways (CP), specifically Canonical Pathways gene sets derived from the KEGG MEDICUS pathway database.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_keggmed_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_keggmed_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_keggmed_lof.csv")

# # Load GMT file as a GeneSetCollection object
# gene_sets_collection <- getGmt(file.path(data_path, "c2.cp.kegg_medicus.v2023.2.Hs.entrez.gmt"))
# 
# # Initialize an empty list to store the results
# gene_sets_list <- list()
# 
# # Iterate over each GeneSet in the GeneSetCollection
# for (setName in names(gene_sets_collection)) {
#   # Extract the GeneSet by its name
#   geneSet <- gene_sets_collection[[setName]]
# 
#   # Convert Entrez IDs to Ensembl IDs for this GeneSet
#   ensembl_ids <- ENTREZtoENSEMBL(geneIds(geneSet))
# 
#   # Ensure the result is a character vector of Ensembl Gene IDs
#   ensembl_ids_vector <- as.character(ensembl_ids)
# 
#   # Store the converted IDs in the list, associated with the setName
#   gene_sets_list[[setName]] <- ensembl_ids_vector
# }
# 
# # Perform FORA analysis for KEGG_MEDICUS Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, gene_sets_list=gene_sets_list)

# Load fora results
fora_results_keggmed_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_keggmed_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_keggmed_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top C2:CP:KEGG_MEDICUS across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_keggmed_all_plot_path <- file.path(figures_gor_path,
                              "mm909_keggmed_all_plot.png")

mm909_keggmed_all_plot <- plot_top5(response_df, fora_results_keggmed_all_df, mm909_keggmed_all_plot_path, "FORA with top C2:CP:KEGG_MEDICUS across samples (All mutations)", "Top C2:CP:KEGG_MEDICUS Gene Sets")

print(mm909_keggmed_all_plot)
```

### Visualization: Top C2:CP:KEGG_MEDICUS across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=20}
# Define output path
mm909_keggmed_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_keggmed_nonsyn_plot.png")

mm909_keggmed_nonsyn_plot <- plot_top5(response_df, fora_results_keggmed_nonsyn_df, mm909_keggmed_nonsyn_plot_path, "FORA with top C2:CP:KEGG_MEDICUS across samples (Nonsyn mutations)", "Top C2:CP:KEGG_MEDICUS Gene Sets")

print(mm909_keggmed_nonsyn_plot)
```

### Visualization: Top C2:CP:KEGG_MEDICUS across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=20}
# Define output path
mm909_keggmed_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_keggmed_lof_plot.png")

mm909_keggmed_lof_plot <- plot_top5(response_df, fora_results_keggmed_lof_df, mm909_keggmed_lof_plot_path, "FORA with top C2:CP:KEGG_MEDICUS across samples (LoF mutations)", "Top C2:CP:KEGG_MEDICUS Gene Sets")

print(mm909_keggmed_lof_plot)
```

## Ninth analysis using C2:CP:REACTOME

Gene sets in this collection are curated from various sources, including online pathway databases and the biomedical literature. Many sets are also contributed by individual domain experts. Here, we focus on Canonical pathways (CP), specifically Canonical Pathways gene sets derived from the Reactome pathway database.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_reactome_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_reactome_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_reactome_lof.csv")

# # Perform FORA analysis for C2_CGP Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, category="C2", subcategory="CP:REACTOME")

# Load fora results
fora_results_reactome_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_reactome_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_reactome_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top C2:CP:REACTOME across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_reactome_all_plot_path <- file.path(figures_gor_path,
                              "mm909_reactome_all_plot.png")

mm909_reactome_all_plot <- plot_top5(response_df, fora_results_reactome_all_df, mm909_reactome_all_plot_path, "FORA with top C2:CP:REACTOME across samples (All mutations)", "Top C2:CP:REACTOME Gene Sets")

print(mm909_reactome_all_plot)
```

### Visualization: Top C2:CP:REACTOME across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_reactome_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_reactome_nonsyn_plot.png")

mm909_reactome_nonsyn_plot <- plot_top5(response_df, fora_results_reactome_nonsyn_df, mm909_reactome_nonsyn_plot_path, "FORA with top C2:CP:REACTOME across samples (Nonsyn mutations)", "Top C2:CP:REACTOME Gene Sets")

print(mm909_reactome_nonsyn_plot)
```

### Visualization: Top C2:CP:REACTOME across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=15}
# Define output path
mm909_reactome_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_reactome_lof_plot.png")

mm909_reactome_lof_plot <- plot_top5(response_df, fora_results_reactome_lof_df, mm909_reactome_lof_plot_path, "FORA with top C2:CP:REACTOME across samples (LoF mutations)", "Top C2:CP:REACTOME Gene Sets")

print(mm909_reactome_lof_plot)
```

## Tenth analysis using MSigDB filter of "Proteasome"

I searched for the term "proteasome" in MSigDB and downloaded a GMT file with all the hits (Gene Sets). Number of GS: 56.

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_proteasome_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_proteasome_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_proteasome_lof.csv")

# # Load GMT file as a GeneSetCollection object
# gene_sets_collection <- getGmt(file.path(data_path, "genesets_proteasome.v2023.2.Hs.gmt"))
# 
# # Initialize an empty list to store the results
# gene_sets_list <- list()
# 
# # Iterate over each GeneSet in the GeneSetCollection
# for (setName in names(gene_sets_collection)) {
#   # Extract the GeneSet by its name
#   geneSet <- gene_sets_collection[[setName]]
# 
#   # Convert Hugo Symbols to Ensembl IDs for this GeneSet
#   ensembl_ids <- SYMBOLtoENSEMBL(geneIds(geneSet))
# 
#   # Ensure the result is a character vector of Ensembl Gene IDs
#   ensembl_ids_vector <- as.character(ensembl_ids)
# 
#   # Store the converted IDs in the list, associated with the setName
#   gene_sets_list[[setName]] <- ensembl_ids_vector
# }
# 
# # Perform FORA analysis for PROTEASOME Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, gene_sets_list=gene_sets_list)

# Load fora results
fora_results_proteasome_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_proteasome_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_proteasome_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]
```

### Visualization: Top PROTEASOME GS across samples stratified by response (All mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
# Define output path
mm909_proteasome_all_plot_path <- file.path(figures_gor_path,
                              "mm909_proteasome_all_plot.png")

mm909_proteasome_all_plot <- plot_top5(response_df, fora_results_proteasome_all_df, mm909_proteasome_all_plot_path, "FORA with top PROTEASOME GS across samples (All mutations)", "Top PROTEASOME Gene Sets")

print(mm909_proteasome_all_plot)
```

### Visualization: Top PROTEASOME GS across samples stratified by response (Nonsyn mutations)

```{r, message=FALSE, warning=FALSE, fig.width=12, fig.height=8}
# Define output path
mm909_proteasome_nonsyn_plot_path <- file.path(figures_gor_path,
                              "mm909_proteasome_nonsyn_plot.png")

mm909_proteasome_nonsyn_plot <- plot_top5(response_df, fora_results_proteasome_nonsyn_df, mm909_proteasome_nonsyn_plot_path, "FORA with top PROTEASOME GS across samples (Nonsyn mutations)", "Top PROTEASOME Gene Sets")

print(mm909_proteasome_nonsyn_plot)
```

### Visualization: Top PROTEASOME GS across samples stratified by response (LoF mutations)

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=7}
# Define output path
mm909_proteasome_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_proteasome_lof_plot.png")

mm909_proteasome_lof_plot <- plot_top5(response_df, fora_results_proteasome_lof_df, mm909_proteasome_lof_plot_path, "FORA with top PROTEASOME GS across samples (LoF mutations)", "Top PROTEASOME Gene Sets")

print(mm909_proteasome_lof_plot)
```

### Visualization: KEGG_PROTEASOME GS across samples stratified by response (LoF mutations)

Filter specifically for KEGG_PROTEASOME gene set.

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}
# Define output path
mm909_keggproteasome_lof_plot_path <- file.path(figures_gor_path,
                              "mm909_proteasome_lof_plot.png")

geneset = c("KEGG_PROTEASOME")

mm909_keggproteasome_lof_plot <- plot_geneset(geneset, response_df, fora_results_proteasome_lof_df, mm909_keggproteasome_lof_plot_path, "FORA with KEGG_PROTEASOME GS across samples (LoF mutations)", "KEGG_PROTEASOME Gene Set")

print(mm909_keggproteasome_lof_plot)
```

# Combined interesting Gene Sets visualizations

I queried MSigDB with 45 interesting Gene Sets that I extracted from the previous exploratory analysis. For some reason, I could not find the GRAESSMANN_APOPTOSIS_BY_DOXORUBICIN_DN and YOSHIMURA_MAPK8_TARGETS_UP for Human when doing this, although I got them when I did the C2:CGP analysis (I always use human in the function). So, I ended up with 43. Here I only present plots containing LoF mutations.

## FORA using a custom GMT file from MSigDB

```{r, message=FALSE, warning=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "fora_results_combinedGS_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "fora_results_combinedGS_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "fora_results_combinedGS_lof.csv")

# Load GMT file as a GeneSetCollection object
gene_sets_collection <- getGmt(file.path(data_path, "genesets_combined.v2023.2.Hs.gmt"))

# Initialize an empty list to store the results
gene_sets_list <- list()

# Iterate over each GeneSet in the GeneSetCollection
for (setName in names(gene_sets_collection)) {
  # Extract the GeneSet by its name
  geneSet <- gene_sets_collection[[setName]]

  # Convert Hugo Symbols to Ensembl IDs for this GeneSet
  ensembl_ids <- SYMBOLtoENSEMBL(geneIds(geneSet))

  # Ensure the result is a character vector of Ensembl Gene IDs
  ensembl_ids_vector <- as.character(ensembl_ids)

  # Store the converted IDs in the list, associated with the setName
  gene_sets_list[[setName]] <- ensembl_ids_vector
}

# # Perform FORA analysis for combined Gene Sets
# perform_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, gene_sets_list=gene_sets_list)

# Load fora results
fora_results_combinedGS_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
fora_results_combinedGS_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
fora_results_combinedGS_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]

# Get all unique sample_id values
unique_sample_ids <- unique(fora_results_combinedGS_lof_df$sample_id)
```

### Visualization: Bar plot per patient

This does not have the filter p.adjust < 0.05.

```{r, message=FALSE, warning=FALSE, fig.width=30.8, fig.height=15}
# Loop through each sample_id
for(sample_id in unique_sample_ids) {
  # Define the output plot path dynamically based on the current sample_id
  plot_path <- file.path(figures_gor_path, paste0(sample_id, "_combinedGS_lof_barplot.png"))
  
  # Get response group
  response <- response_df %>% 
    dplyr::filter(sample_id == !!sample_id) %>% 
    dplyr::select(patient_response)
  response <- as.character(response)
  
  # Generate the plot title dynamically
  plot_title <- paste("FORA with combined GS for sample", sample_id, "-", response, "(LoF mutations)")
  
  # Call the plotting function with the current sample_id
  combinedGS_lof_plot <- plot_combinedGS(sample_id, response_df, fora_results_combinedGS_lof_df, plot_path, plot_title, 105, 30.8)
  
  # Print
  print(combinedGS_lof_plot)
}
```

### Visualization: Bubble plot per Gene Set

This does not have the filter p.adjust < 0.05.

```{r, message=FALSE, warning=FALSE, fig.width=15, fig.height=5}
# Loop through each geneset
for(geneset in names(gene_sets_list)) {
  # Define the output plot path dynamically based on the current geneset
  plot_path <- file.path(figures_gor_path, paste0(geneset, "_combinedGS_lof_bubbleplot.png"))
  
  # Generate the plot title dynamically
  plot_title <- paste("FORA with", geneset, "GS per sample stratified by response (LoF mutations)")
  y_label <- NULL
  
  # Call the plotting function with the current geneset
  combinedGS_lof_plot <- plot_geneset(geneset, response_df, fora_results_combinedGS_lof_df, plot_path, plot_title, y_label)
  
  # Print
  print(combinedGS_lof_plot)
}
```

### Visualization: Box plot per 10 Gene Sets

This does not have the filter p.adjust < 0.05.

```{r, message=FALSE, warning=FALSE, fig.width=20, fig.height=12}
# Calculate the sequence of indices in steps of 10
indices <- seq(from = 1, to = length(gene_sets_list), by = 10)

# Loop through the indices
global_idx <- 1
for (i in indices) {
  # Select a group of 10 gene sets or the remaining ones if less than 10
  geneset_group <- names(gene_sets_list)[i:min(i+9, length(gene_sets_list))]
  
  # Define the output plot path dynamically based on the current index
  plot_path <- file.path(figures_gor_path, paste0(global_idx, "_combinedGS_lof_boxplot.png"))
  
  # Call the plotting function with the current geneset group
  combinedGS_lof_plot <- boxplot_genesets(geneset_group, response_df, fora_results_combinedGS_lof_df, plot_path, 
                                      paste("FORA with GS", i, "to", min(i+9, length(gene_sets_list)), 
                                            "stratified by response (LoF mutations)"), 110)
  
  # Print
  print(combinedGS_lof_plot)
  
  # Update global index
  global_idx = global_idx + 1
}
```

### Visualization: Global heat map

This does not have the filter p.adjust < 0.05.

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=10}
plot_path <- file.path(figures_gor_path, "mm909_combinedGS_lof_heat_map.png")
combinedGS_lof_plot <- heat_map_genesets(response_df, fora_results_combinedGS_lof_df, plot_path, 
                                    "Heat Map of combined Gene Sets per patient")

# Print
print(combinedGS_lof_plot)
```

## Using clusterProfiler

```{r, warning=FALSE, message=FALSE}
# Define paths
fora_results_all_csv_path <- file.path(results_gor_path, "cP_fora_results_combinedGS_all.csv")
fora_results_nonsyn_csv_path <- file.path(results_gor_path, "cP_fora_results_combinedGS_nonsyn.csv")
fora_results_lof_csv_path <- file.path(results_gor_path, "cP_fora_results_combinedGS_lof.csv")

# Load GMT file as a GeneSetCollection object
gene_sets_collection <- getGmt(file.path(data_path, "genesets_combined.v2023.2.Hs.gmt"))

# Create an empty dataframe
gene_sets_df <- tibble(gene_set_name = character(), gene_hs = character())

# Iterate over each gene set in the collection
for (gene_set_name in names(gene_sets_collection)) {
  # Access the S4 object
  gene_set <- gene_sets_collection[[gene_set_name]]
  set_name <- gene_set@setName
  gene_hs <- gene_set@geneIds
  
  # Create a temporary dataframe
  temp_df <- tibble(
    gene_set_name = set_name,
    gene_hs = gene_hs
  )
  
  # Bind the temp dataframe to the main dataframe
  gene_sets_df <- bind_rows(gene_sets_df, temp_df)
  
  # Convert Hugo Symbols to Ensembl IDs for this GeneSet
  gene_sets_df <- gene_sets_df %>% 
    dplyr::mutate(ensembl_id = SYMBOLtoENSEMBL(gene_hs))
}

# # Perform FORA analysis for combined Gene Sets
# results_list <- perform_cP_fora_analysis(sample_id_list, list_of_all_mutations, list_of_nonsyn_mutations, list_of_lof_mutations, fora_results_all_csv_path, fora_results_nonsyn_csv_path, fora_results_lof_csv_path, gene_sets_df=gene_sets_df)

# Load fora results
cP_fora_results_combinedGS_all_df <- read.csv(fora_results_all_csv_path, header = TRUE)[ , -1]
cP_fora_results_combinedGS_nonsyn_df <- read.csv(fora_results_nonsyn_csv_path, header = TRUE)[ , -1]
cP_fora_results_combinedGS_lof_df <- read.csv(fora_results_lof_csv_path, header = TRUE)[ , -1]

# Get all unique sample_id values
unique_sample_ids <- unique(cP_fora_results_combinedGS_lof_df$sample_id)
```

```{r, warning=FALSE, message=FALSE}
for(sample_id in unique_sample_ids) {
  # Define the output plot path dynamically based on the current sample_id
  plot_path <- file.path(figures_gor_path, paste0(sample_id, "_combinedGS_lof_upsetplot.png"))
  tryCatch({
    # Attempt to generate and print the plot
    upset_plot <- enrichplot::upsetplot(results_list$lof[[sample_id]])
    
    # Print
    print(upset_plot)
    
    # Save plot
    ggsave(plot_path, upset_plot, width = 10, height = 10, dpi = 300)
    
  }, error = function(e) {
    # If an error occurs, print the error message
    cat("An error occurred with sample", sample_id, ": ", e$message, "\n")
  })
}
```

```{r}
# Assuming you have two gene sets identified from the upset plot you wish to compare
gene_set_ids <- c("GOBP_PROTEIN_K48_LINKED_DEUBIQUITINATION", "GOBP_PROTEIN_MODIFICATION_BY_SMALL_PROTEIN_REMOVAL", "GOMF_DEUBIQUITINASE_ACTIVITY", "GOMF_UBIQUITIN_LIKE_PROTEIN_PEPTIDASE_ACTIVITY", "REACTOME_DEUBIQUITINATION")

# Extract gene members for each set from the enricher results
# This step may vary depending on the exact structure of your enricher results
gene_members <- lapply(gene_set_ids, function(set_id) {
  genes_in_set <- results_list$lof[["MM909_14"]]@result$geneID[which(results_list$lof[["MM909_14"]]@result$ID == set_id)]
  unlist(strsplit(genes_in_set, "/"))  # Assuming genes are separated by "/"
})

# Now, gene_members is a list where each element contains the genes for a gene set

# Find shared genes between the sets
shared_genes <- Reduce(intersect, gene_members)

list_of_lof_mutations[["MM909_14"]] %>% 
  dplyr::filter(ensembl_id %in% shared_genes)
```

I find it weird that clusterProfiler finds much less enriched GS than the other approach. Still, I investigated the intersection between some GS of patient **MM909_14**. For GS: ; I saw that they shared only 1 gene: **ENSG00000169914 (OTUD3)**. This gene is protein coding, has 3 transcripts and belongs to the **OTU (Ovarian Tumor) family deubiquitinases**. These proteins are a group of enzymes that belong to the cysteine protease class, specifically involved in the process of deubiquitination. OTU deubiquitinases play significant roles in regulating various cellular processes by reversing the ubiquitination of proteins, thus affecting protein stability, localization, and function. Through their action, OTU family deubiquitinases are involved in regulating cell cycle progression, DNA repair, signal transduction, and immune responses, making them important for maintaining cellular homeostasis and implicated in the pathogenesis of diseases, including cancer and inflammation. Link: https://grch37.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000169914;r=1:19882395-19912945

```{r}
# Assuming you have two gene sets identified from the upset plot you wish to compare
gene_set_ids <- c("GOBP_POSITIVE_REGULATION_OF_PROTEIN_UBIQUITINATION", "HALLMARK_P53_PATHWAY")

# Extract gene members for each set from the enricher results
# This step may vary depending on the exact structure of your enricher results
gene_members <- lapply(gene_set_ids, function(set_id) {
  genes_in_set <- results_list$lof[["MM909_14"]]@result$geneID[which(results_list$lof[["MM909_14"]]@result$ID == set_id)]
  unlist(strsplit(genes_in_set, "/"))  # Assuming genes are separated by "/"
})

# Now, gene_members is a list where each element contains the genes for a gene set

# Find shared genes between the sets
shared_genes <- Reduce(intersect, gene_members)

list_of_lof_mutations[["MM909_14"]] %>% 
  dplyr::filter(ensembl_id %in% shared_genes)
```
I investigated the intersection between some GS of patient **MM909_14**. For GS: ; I saw that they shared only 1 gene: **ENSG00000171720 (HDAC3)**. This gene is protein coding, has 16 transcripts and belongs to the **Class I histone deacetylases (HDACs).**. These proteins are a subgroup of the HDAC family, enzymes that play a crucial role in the regulation of gene expression by removing acetyl groups from lysine residues on histone proteins. This deacetylation leads to a more compact chromatin structure, resulting in transcriptional repression. Class I HDACs, which include HDAC1, HDAC2, HDAC3, and HDAC8, are primarily located in the nucleus and are involved in various cellular processes such as cell cycle progression, differentiation, and apoptosis. Their activity is tightly regulated, and dysregulation of Class I HDACs has been linked to the development and progression of several diseases, including cancer, where they can influence tumor suppressor genes and oncogenes.
Link: https://grch37.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000171720;r=5:141000443-141016437

# Session Info

```{r}
sessionInfo()
```
