---
title: "Impact of dysfunctional ubiquitination in response to cancer immunotherapy"
subtitle: <center> Somatic mutations analysis - Protein sequence FASTA file generation </center>
author: "Eric Bautista Farrerons (s212514)"
date: '`r paste("First created on April 2024. Updated on ", format(Sys.Date(), "%d %B %Y"))`'
output:
  html_document:
    css: style.css
    code_folding: hide
    fig_caption: yes
    toc: yes
    toc_float:
      collapsed: true
      smooth_scroll: true
    toc_depth: 4
    number_sections: true
---

# Data pre-processing
### Load data

This document is to get the protein sequences of the ubiquitin-related genes that have a LoF mutation from all the samples and create a multifasta file.

```{r, message=FALSE}
### Load required libraries
library(readxl)
library(tidyverse)
library(purrr)
library(dplyr)
library(knitr)
library(DT)
library(httr)
library(jsonlite)
library(stringr)
library(biomaRt)
library(GSEABase)
library(HGNChelper)
library(mygene)
library(GO.db)
library(AnnotationDbi)
library(VennDiagram)

### Load functions
source(file = "../02_functions.R")

### Define paths
current_dir <- getwd()
data_path <- file.path(current_dir,
                       "../../data/_raw")
results_pc_path <- file.path(current_dir,
                              "../../results/patient_characteristics/output")
results_fip_path <- file.path(current_dir,
                              "../../results/somatic_mutation_analysis/functional_impact_prediction/output")
results_ubgenesets_path <- file.path(current_dir, 
                                 "../../results/somatic_mutation_analysis/ubiquitin_custom_GS/output")
results_ubfora_path <- file.path(current_dir, 
                                 "../../results/somatic_mutation_analysis/ubiquitin_fora/output")
results_protein_seqs_path <- file.path(current_dir, 
                                 "../../results/somatic_mutation_analysis/protein_seqs/output")

# Create the folder if it does not exist
if(!file.exists(results_protein_seqs_path)) {
  dir.create(results_protein_seqs_path, recursive = TRUE)
}

### Read data & wrangle
patients_ub_lof_mut_path <- file.path(results_ubfora_path, "lof_patients_ub_mutations.csv") # LoF patients mutations df
patients_ub_lof_mut_df <- read.csv(patients_ub_lof_mut_path, header = TRUE)[ , -1]

# Extract ensembl_ids to do the left_join
g_ensembl_df <- patients_ub_lof_mut_df %>% 
  dplyr::select(ensembl_id) %>% 
  dplyr::distinct()

lof_df_path <- file.path(results_fip_path, "lof_mutations.csv") # LoF mutations df
lof_df_annotated <- read.csv(lof_df_path, header = TRUE)[ , -1]

# Left join Ensembl Protein ID
patients_ub_prot_lof_mut_df <- g_ensembl_df %>% 
  dplyr::left_join(lof_df_annotated, by=c("ensembl_id")) %>% 
  dplyr::rename(p_ensembl_id = ENSP) %>% 
  dplyr::select(sample_id, ensembl_id, p_ensembl_id, SWISSPROT, Location, Allele, Liberal_consequence) %>% 
  dplyr::distinct()

# Save to file
patients_ub_prot_lof_mut_path <- file.path(results_protein_seqs_path, 
                      "prot_lof_patients_ub_mutations.csv")
write.csv(patients_ub_prot_lof_mut_df, file = patients_ub_prot_lof_mut_path, row.names = TRUE)

# Extract unique protein ensembl ids
p_ensembl_ids <- unique(patients_ub_prot_lof_mut_df$p_ensembl_id) # Gives 349 protein_ids for 106 genes
swiss_prot_ids <- unique(patients_ub_prot_lof_mut_df$SWISSPROT) # Missing a few (101 swissprot_ids vs 106 genes)
```

# Getting Protein sequences using biomaRt (grch37 assembly)

```{r, message=FALSE, warning=FALSE}
ensembl = biomaRt::useMart(biomart="ENSEMBL_MART_ENSEMBL",
                  host="https://grch37.ensembl.org",
                  path="/biomart/martservice",
                  dataset="hsapiens_gene_ensembl")

# Retrieve protein sequences
protein_sequences <- biomaRt::getSequence(id = p_ensembl_ids, 
                                          type = "ensembl_peptide_id", 
                                          seqType = "peptide", 
                                          mart = ensembl)
```

# Create the Multifasta File

```{r, message=FALSE, warning=FALSE}
# Define a function to format as FASTA
format_fasta <- function(id, sequence) {
  paste0(">", id, "\n", sequence)
}

# Apply the function and combine into a single string
fasta_content <- sapply(1:nrow(protein_sequences), function(i) {
  format_fasta(protein_sequences$ensembl_peptide_id[i], protein_sequences$peptide[i])
})

# Write to a FASTA file
multifasta_path <- file.path(results_protein_seqs_path, 
                                 "multifasta_ub_lof_mut.fasta")
writeLines(unlist(fasta_content), multifasta_path)
```

# Split multifasta in files of 100 seqs for PFAM analysis

```{r, message=FALSE, warning=FALSE}
# Determine the number of files needed (each containing up to 100 sequences)
num_files <- ceiling(length(fasta_content) / 100)

# Write each chunk to a separate file
for (file_number in 1:num_files) {
  # Calculate the indices for slicing the fasta_content
  start_index <- (file_number - 1) * 100 + 1
  end_index <- min(file_number * 100, length(fasta_content))
  
  # Subset the fasta content for this file
  fasta_subset <- fasta_content[start_index:end_index]
  
  # Define the path for this particular chunk
  file_path <- file.path(results_protein_seqs_path, 
                         sprintf("multifasta_ub_lof_mut_part_%03d.fasta", file_number))
  
  # Write to a FASTA file
  writeLines(unlist(fasta_subset), file_path)
}
```

# Session Info

```{r}
sessionInfo()
```
