---
title: "Impact of dysfunctional ubiquitination in response to cancer immunotherapy"
subtitle: <center> Somatic mutations analysis </center>
author: "Eric Bautista Farrerons (s212514)"
date: '`r paste("First created on January 2024. Updated on ", format(Sys.Date(), "%d %B %Y"))`'
output:
  html_document:
    css: style.css
    code_folding: hide
    fig_caption: yes
    toc: yes
    toc_float:
      collapsed: true
      smooth_scroll: true
    toc_depth: 4
    number_sections: true
---

```{r, message=FALSE}
### Load required libraries
library(readxl)
library(tidyverse)
library(dplyr)
library(knitr)
library(DT)
library(httr)
library(jsonlite)
library(stringr)

### Load functions
source(file = "../R/02_functions.R")

### Define paths
current_dir <- getwd()

# MAF-like file
maf_data <- "../data/_raw/41467_2017_1460_MOESM6_ESM_somatic_mutations.xlsx"
maf_path <- file.path(current_dir, 
                      maf_data)

### Read data
maf_df <- read_excel(maf_path,
                        skip=1,
                        col_names=TRUE)

### Add a sample ID column
maf_df <- maf_df %>%
  mutate(sample_id = sub("_[12]$", "", tumor_name)) %>% 
  dplyr::select(sample_id, everything()) %>% 
  arrange(sample_id)
```

## See mutations per patient

```{r}
unique_tumor_counts <- maf_df %>%
  dplyr::count(sample_id, 
        sort = TRUE)  # Count unique values and sort

# Display the result
datatable(unique_tumor_counts, 
          extensions = 'Buttons', 
          options = list(
            dom = 'Bfrtip',
            buttons = c('copy', 'excel', 'csv'),
            scrollX=TRUE,
            pageLength=10,
            columnDefs = list(list(
              targets = "_all",
              render = JS(
                "function(data, type, row, meta) {",
                "  return data === null ? 'NA' : data;",
                "}"
              )
            ))
          ),
          caption = "Mutation counts per patient"
        )
```

## Exploration

```{r, warning=FALSE}
# Filter rows where Entrez_Gene_Id is 0
missing_entrez_df <- maf_df %>%
  filter(Entrez_Gene_Id == 0)

# Filter rows where Hugo_Symbol is Unknown
missing_hugo_df <- maf_df %>%
  filter(Hugo_Symbol == 'Unknown')

n_rows_1 <- nrow(missing_entrez_df)
n_rows_2 <- nrow(missing_hugo_df)

print(paste("Number of genomic regions missing an Entrez ID:", n_rows_1))
print(paste("Number of genomic regions missing Hugo Symbol:", n_rows_2))

# How many different types of mutations has the dataset?
unique_values <- unique(maf_df$Variant_Classification)
print(paste("The mutation categories in the data:", toString(unique_values)))

# For selected nonsynonymous mutations based on literature
selected_mutations <- c(unique_values[1], unique_values[3], unique_values[7], unique_values[10], unique_values[12])
print(paste("The mutation categories I will use as Nonsynonymous based on literature:", toString(selected_mutations)))
```

## Get a list of mutations for each patient

```{r}
# Splitting the dataframe into a list based on Sample.ID
list_of_all_mutations <- split(maf_df, maf_df$sample_id)
```

## Get a list of Nonsynonymous mutations per patient

```{r}
nonsyn_maf_df <- maf_df %>%
  filter(Variant_Classification %in% c("Missense_Mutation", "Nonsense_Mutation", "Nonstop_Mutation", "Start_Codon_SNP", "Splice_Site")) %>% 
  arrange(sample_id)

list_of_nonsyn_mutations <- split(nonsyn_maf_df, nonsyn_maf_df$sample_id)
```

## Export info to use Ensembl VEP (web version)

```{r}
### Extract nonsyn info in correct format for VEP input
api_input <- nonsyn_maf_df %>% 
  dplyr::select(Chromosome,
         Start_position,
         End_position,
         ref_allele,
         alt_allele,
         Strand,
         sample_id) %>% 
  dplyr::mutate(Allele_ref_alt = paste(ref_allele, 
                                alt_allele, 
                                sep = "/")) %>% 
  dplyr::mutate(API_info = paste(Chromosome,
                          Start_position,
                          End_position,
                          Allele_ref_alt,
                          Strand,
                          sample_id,
                          sep = " "))

### Export the info to a file and use the Web VEP
vep_input_data <- "../data/vep_input_nonsyn_grch37.txt"
vep_file_path <- file.path(current_dir, 
                      vep_input_data)
write.table(api_input$API_info, file = vep_file_path, row.names = FALSE, col.names = FALSE, quote = FALSE)
```

## Run VEP and explore output

```{r, warning=FALSE}
### Define path
vep_out_data <- "../data/vep_output_nonsyn_grch37_03.txt"
vep_path <- file.path(current_dir,
                      vep_out_data)

### Read data
vep_df <- read.csv(vep_path, header = TRUE, sep = "\t")

### Clean and wrangle
vep_df_clean <- vep_df %>% 
  mutate(across(everything(), ~na_if(., "-"))) %>% 
  select(where(~any(!is.na(.)))) %>% 
  rename(sample_id = X.Uploaded_variation)

print(paste("The columns that do not have any information are:", toString(setdiff(colnames(vep_df), colnames(vep_df_clean)))))
```

## Get a list of LoF mutations per patient

```{r}
lof_vep_df <- vep_df_clean %>%
  mutate(SIFT_cleaned = str_extract(SIFT, "^[^(]+"), 
         PolyPhen_cleaned = str_extract(PolyPhen, "^[^(]+")) %>%
  filter(SIFT_cleaned %in% c("deleterious", "deleterious_low_confidence") | 
         PolyPhen_cleaned %in% c("probably_damaging", "possibly_damaging")) %>% 
  arrange(sample_id)

list_of_lof_mutations <- split(lof_vep_df, lof_vep_df$sample_id)
```

## Session Info

```{r}
sessionInfo()
```
